### PR TITLE
WT-3913 Enhance cursor operations to account for prepare : WIP

### PR DESCRIPTION
**Pull request by bvpvamsikrishna**
Work in progress.
Modify to return WT_ROLLBACK error for updated operations when encounter a prepared updated.

### CODE REVIEW

File: src/include/txn.i [hunk 1]
```diff
@@ -480,31 +485,59 @@ __txn_visible_id(WT_SESSION_IMPL *session, uint64_t id)
  * __wt_txn_visible --
  *	Can the current transaction see the given ID / timestamp?
  */
-static inline bool
+static inline int
 __wt_txn_visible(
-    WT_SESSION_IMPL *session, uint64_t id, const wt_timestamp_t *timestamp)
+    WT_SESSION_IMPL *session, uint64_t id, const wt_timestamp_t *timestamp,
+    WT_UPDATE *upd)
```
Comment (michaelcahill) [inline src/include/txn.i]: It seems wrong to me that we now have `__wt_txn_visible` taking an update.  I expected any logic that required an update to stay in `__wt_txn_upd_visible` and this function just deals with the visibility of a transaction ID.

File: src/include/txn.i [hunk 2]
```diff
@@ -480,31 +485,59 @@ __txn_visible_id(WT_SESSION_IMPL *session, uint64_t id)
...
+				__wt_yield();
+			} while (FLD_ISSET(upd->state, WT_UPDATE_STATE_LOCKED));
+			/* FALLTHROUGH */
+		case WT_UPDATE_STATE_READY:
+			return (!F_ISSET(txn, WT_TXN_HAS_TS_READ) ?
+			    WT_VISIBLE_TYPE_TRUE :
+			    (__wt_timestamp_cmp(timestamp,
+				&txn->read_timestamp) <= 0 ?
+				WT_VISIBLE_TYPE_TRUE : WT_VISIBLE_TYPE_FALSE));
+		}
```
Comment (michaelcahill) [inline src/include/txn.i]: I think this logic is basically correct, just not where I expected it.

File: src/include/txn.i [hunk 3]
```diff
@@ -514,8 +547,21 @@ __wt_txn_visible(
 static inline bool
 __wt_txn_upd_visible(WT_SESSION_IMPL *session, WT_UPDATE *upd)
 {
-	return (__wt_txn_visible(session,
-	    upd->txnid, WT_TIMESTAMP_NULL(&upd->timestamp)));
+	WT_VISIBLE_TYPE visible_type;
+
+	visible_type = __wt_txn_visible(session, upd->txnid,
+	    WT_TIMESTAMP_NULL(&upd->timestamp), upd);
```
Comment (michaelcahill) [inline src/include/txn.i]: What I expected was for `__wt_txn_visible` to continue a boolean.

Then wait if `upd->state == LOCKED`.  Then save `upd->state`, call `__wt_txn_visible` and make sure the state doesn't change (loop if it does).

**Comment by bvpvamsikrishna** [comment]
Will raise a new pull request for this ticket.

**Pull request closed without merging**